cmd/kubectl-gadget/utils/trace.go

```diff
@@ -374,59 +377,101 @@ func SetTraceOperation(traceID string, operation string) error {
-func waitForTraceState(traceID string, expectedState string) (*gadgetv1alpha1.TraceList, error) {
-	start := time.Now()
+// getTraceWatcher returns a watcher on trace(s) whom ID was given as parameter.
+// This watcher can then be used to wait on State.Output modification.
+func getTraceWatcher(traceID string) (watch.Interface, error) {
+	traceClient, err := getTraceClient()
+	if err != nil {
+		return watch.NewEmptyWatch(), err
+	}
 
-RetryLoop:
-	for {
-		successNodeCount := 0
-		timeout := time.Since(start) > traceTimeout
-		nodeErrors := make(map[string]string)
-		nodeWarnings := make(map[string]string)
+	var watchOptions = metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", GLOBAL_TRACE_ID, traceID),
+	}
 
-		traces, err := getTraceListFromID(traceID)
-		if err != nil {
-			return nil, err
+	watcher, err := traceClient.GadgetV1alpha1().Traces("gadget").Watch(context.TODO(), watchOptions)
+	if err != nil {
+		return watch.NewEmptyWatch(), err
+	}
+
+	return watcher, nil
+}
+
+// waitForTraceState wait for traces whom ID is given as parameter to be in the
+// expected state.
+func waitForTraceState(traceID string, expectedState string) (gadgetv1alpha1.TraceList, error) {
+	var returnedTraces gadgetv1alpha1.TraceList
+
+	traces, err := getTraceListFromID(traceID)
+	if err != nil {
+		return gadgetv1alpha1.TraceList{}, err
+	}
+
+	watchedTracesNumber := 0
+	tracesNumber := len(traces.Items)
+
+	watcher, err := getTraceWatcher(traceID)
+	if err != nil {
+		return gadgetv1alpha1.TraceList{}, err
+	}
+
+	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), traceTimeout)
+	_, err = watchtools.UntilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error){
+		// Deal particularly with error.
+		if event.Type == watch.Error {
+			return false, err
 		}
 
-		for _, i := range traces.Items {
-			if i.Status.OperationError != "" {
-				nodeErrors[i.Spec.Node] = i.Status.OperationError
-			} else if i.Status.OperationWarning != "" {
-				nodeWarnings[i.Spec.Node] = i.Status.OperationWarning
-				// TODO(francis) This code will not work if the trace is already in the
-				// expected state, for example if we decide to generate it twice.
-				// We need to add a cookie (and check it) to be sure the trace is ready.
-			} else if i.Status.State == expectedState {
-				successNodeCount++
-			} else {
-				// Consider Trace as timed out if it neither moved the state forward
-				// nor notified of an error or warning within the time window.
-				if timeout {
-					nodeErrors[i.Spec.Node] = fmt.Sprintf("No results received from trace within %v", traceTimeout)
-					continue
-				}
+		// We are only interested in Added and Modified event, as we want
+		// Status.State value to change.
+		// More particularly, we monitor Added event for gadget like
+		// process-collector and Modified for gadget like seccompadvisor.
+		if event.Type != watch.Added && event.Type != watch.Modified {
+			return false, nil
+		}
```

Comment:
Are we really interested on `Added` events? `tracesNumber` is defined so lines above, I think an added tracer will mess up this count.

Reply:
I will check again, but I added `Added` events for gadget like `process-collector` which are never modified.

Reply:
I checked and we are not interested in this type of event, good catch!

Reply:
Hum! I am wrong, we need it for `dns` gadget.

Reply:
Why do we need it for the DNS gadget? At the time this code is run the trace should already exists, shouldn't it?

Reply:
From my test, I only get `watch.Added` when I ran it with `dns`:

```bash
./kubectl-gadget dns -n demo
POD                            TYPE      NAME
event.Type: ADDED
event.Type: ADDED
dns0                           OUTGOING  microsoft.com.demo.svc.cluster.local.
...
```

I think this is due to how the gadget are implemented on the trace side, `dns` has [`Start()`](https://github.com/kinvolk/inspektor-gadget/blob/8877e4b7d94e6e56e19d001ea6df812c41697e72/pkg/gadgets/dns/gadget.go#L111) and [`Stop()`](https://github.com/kinvolk/inspektor-gadget/blob/8877e4b7d94e6e56e19d001ea6df812c41697e72/pkg/gadgets/dns/gadget.go#L237) while `*-collector` have [`Collect()`](https://github.com/kinvolk/inspektor-gadget/blob/8877e4b7d94e6e56e19d001ea6df812c41697e72/pkg/gadgets/process-collector/gadget.go#L64).

Reply:
I performed some tests and I found that in both events are received, hence having only `watch.Added` should be enough. Can you please try it out?

Reply:
I tested removing `event.Type == watch.Modified` with `process-collector` and `biolatency` and I got the following at each run:

```
event.Type: ADDED
event.Type: MODIFIED
event.Type: MODIFIED
timed out waiting for the condition
```

Thus, I think it is better to keep the condition as it is now.

Reply:
After reading a bit on the details I think I now understand what's going on. We aren't specifying the `ResourceVersion` when calling `traceClient.GadgetV1alpha1().Traces("gadget").Watch`, it means the watcher will default to the most recent version of the resource. Quoting the important part of the [docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-watch):

```
To establish initial state, the watch begins with synthetic "Added" events of all resources instances that exist at the starting resource version
```

So those added events we're getting are to get the initial status of the resources. Given that I think you can remove the  call to `getTraceListFromID` and use those added events to get the number of traces we're interested in. I think something like the following could work:

```go
	// set outside the conditional function
	tracesNumber := 0


		// in the conditional function 
		if event.Type == watch.Added {
			// there is another trace to watch
			tracesNumber++

			// check if this trace is already in the status we're looking for
			trace := event.Object.(*gadgetv1alpha1.Trace)
			if trace.Status.State == expectedState {
				watchedTracesNumber++
			}

			return watchedTracesNumber == tracesNumber, nil
		}
```